AtomicFile issues

```

04:25:05 Caused by: java.nio.file.NoSuchFileException: /jenkins/jenkins-work/nodes/yadp877105322/atomic113511861761462995tmp
04:25:05 	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
04:25:05 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
04:25:05 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
04:25:05 	at sun.nio.fs.UnixCopyFile.move(UnixCopyFile.java:409)
04:25:05 	at sun.nio.fs.UnixFileSystemProvider.move(UnixFileSystemProvider.java:262)
04:25:05 	at java.nio.file.Files.move(Files.java:1395)
04:25:05 	at hudson.util.AtomicFileWriter.commit(AtomicFileWriter.java:206)
04:25:05 	at hudson.XmlFile.write(XmlFile.java:198)
04:25:05 	at jenkins.model.Nodes.save(Nodes.java:274)
04:25:05 	at hudson.util.PersistedList.onModified(PersistedList.java:173)
04:25:05 	at hudson.util.PersistedList.replaceBy(PersistedList.java:85)
04:25:05 	at hudson.model.Slave.<init>(Slave.java:198)
04:25:05 	at hudson.slaves.AbstractCloudSlave.<init>(AbstractCloudSlave.java:51)
04:25:05 	at com.github.kostyasha.yad.DockerSlaveSingle.<init>(DockerSlaveSingle.java:53)
04:25:05 	at com.github.kostyasha.yad.DockerSimpleBuildWrapper.setUp(DockerSimpleBuildWrapper.java:94)
04:25:05 	... 14 more
04:25:05 	Suppressed: java.nio.file.NoSuchFileException: /jenkins/jenkins-work/nodes/yadp877105322/atomic113511861761462995tmp -> /jenkins/jenkins-work/nodes/yadp877105322/config.xml
04:25:05 		at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
04:25:05 		at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
04:25:05 		at sun.nio.fs.UnixCopyFile.move(UnixCopyFile.java:396)
04:25:05 		at sun.nio.fs.UnixFileSystemProvider.move(UnixFileSystemProvider.java:262)
04:25:05 		at java.nio.file.Files.move(Files.java:1395)
04:25:05 		at hudson.util.AtomicFileWriter.commit(AtomicFileWriter.java:191)
04:25:05 		... 22 more
....

--- a/yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java
+++ b/yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java
@@ -263,7 +263,30 @@ public class DockerCloud extends AbstractCloud implements Serializable {
         }
 
         final DockerComputerLauncher launcher = computerLauncher.getPreparedLauncher(getDisplayName(), template, ir);
-        return new DockerSlave(slaveName, nodeDescription, launcher, containerId, template, getDisplayName(), id);
+
+        return getDockerSlaveWithRetry(template, id, containerId, nodeDescription, slaveName, launcher);
+    }
+
+    private DockerSlave getDockerSlaveWithRetry(DockerSlaveTemplate template, ProvisioningActivity.Id id,
+                                                String containerId, String nodeDescription, String slaveName,
+                                                DockerComputerLauncher launcher)
+            throws IOException, Descriptor.FormException, InterruptedException {
+        int retries = 6;
+        while (retries >= 0) {
+            try {
+                return new DockerSlave(slaveName, nodeDescription, launcher, containerId, template, getDisplayName(), id);
+            } catch (IOException t) {
+                if (retries <= 0) {
+                    throw t;
+                }
+                LOG.trace("Failed to create DockerSlaveSingle, retrying...", t);
+                Thread.sleep(1000);
+            } finally {
+                retries--;
+            }
+        }
+
+        throw new IllegalStateException("Max creation retries");
     }
 
     /**

--- a/yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerSimpleBuildWrapper.java
+++ b/yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerSimpleBuildWrapper.java
@@ -27,6 +27,7 @@ import javax.annotation.Nonnull;
 import java.io.IOException;
 import java.io.PrintStream;
 
+import static com.github.kostyasha.yad.utils.DockerUtils.jenkinsAddNodeWithRetry;
 import static java.util.Objects.isNull;
 import static org.jenkinsci.plugins.cloudstats.CloudStatistics.ProvisioningListener.get;
 
@@ -88,14 +89,8 @@ public class DockerSimpleBuildWrapper extends SimpleBuildWrapper {
             get().onStarted(activityId);
             final String futureName = "yadp" + activityId.getFingerprint();
 
-            final DockerSlaveSingle slave = new DockerSlaveSingle(futureName,
-                    "Slave for " + run.getFullDisplayName(),
-                    getConfig(),
-                    getConnector(),
-                    activityId
-            );
+            jenkinsAddNodeWithRetry(getDockerSlaveSingleWithRetry(run, activityId, futureName));
 
-            Jenkins.getInstance().addNode(slave);
             final Node futureNode = Jenkins.getInstance().getNode(futureName);
             if (!(futureNode instanceof DockerSlaveSingle)) {
                 logger.println("Can't get node" + futureName);
@@ -115,14 +110,15 @@ public class DockerSimpleBuildWrapper extends SimpleBuildWrapper {
                 computer.setListener(listener);
                 node.setListener(listener);
                 logger.println("Getting launcher...");
-                ((DelegatingComputerLauncher) computer.getLauncher()).getLauncher().launch(computer, listener);
+                ((DelegatingComputerLauncher) computer.getLauncher()).getLauncher()
+                        .launch(computer, listener);
             } catch (Throwable e) {
                 logger.println("Failed to provision");
                 e.printStackTrace(logger);
                 LOG.error("fd", e);
                 CloudStatistics.ProvisioningListener.get().onFailure(node.getId(), e);
                 //terminate will do container cleanups and remove node from jenkins silently
-                slave.terminate();
+                node.terminate();
                 throw e;
             }
 
@@ -136,6 +132,31 @@ public class DockerSimpleBuildWrapper extends SimpleBuildWrapper {
             throw new IOException("failed to run slave", e);
         }
     }
+
+    private DockerSlaveSingle getDockerSlaveSingleWithRetry(Run<?, ?> run, ProvisioningActivity.Id activityId, String futureName)
+            throws IOException, Descriptor.FormException, InterruptedException {
+        int retries = 6;
+        while (retries >= 0) {
+            try {
+                return new DockerSlaveSingle(futureName,
+                        "Slave for " + run.getFullDisplayName(),
+                        getConfig(),
+                        getConnector(),
+                        activityId);
+            } catch (IOException t) {
+                if (retries <= 0) {
+                    throw t;
+                }
+                LOG.trace("Failed to create DockerSlaveSingle, retrying...", t);
+                Thread.sleep(1000);
+            } finally {
+                retries--;
+            }
+        }
+
+        throw new IllegalStateException("Max creation retries");
+    }
+
 
     /**
      * Terminates slave for specified slaveName. Works only with {@link DockerSlaveSingle}.

--- a/yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/utils/DockerUtils.java
+++ b/yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/utils/DockerUtils.java
@@ -1,0 +1,45 @@
+package com.github.kostyasha.yad.utils;
+
+import com.github.kostyasha.yad.DockerSlaveSingle;
+import jenkins.model.Jenkins;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+import static java.util.Objects.nonNull;
+
+public class DockerUtils {
+    private static final Logger LOG = LoggerFactory.getLogger(DockerUtils.class);
+
+    private DockerUtils() {
+    }
+
+    public static void jenkinsAddNodeWithRetry(DockerSlaveSingle slave) throws IOException,
+            InterruptedException {
+        IOException lastException = null;
+        int retries = 6;
+
+        while (retries >= 0) {
+            try {
+                Jenkins.getInstance().addNode(slave);
+            } catch (IOException ex) {
+                lastException = ex;
+                if (retries <= 0) {
+                    throw ex;
+                }
+                LOG.trace("Failed to add DockerSlaveSingle node to Jenkins, retrying...", ex);
+                Thread.sleep(1000);
+            } finally {
+                retries--;
+            }
+        }
+        if (nonNull(lastException)) {
+            throw lastException;
+        }
+
+        throw new IllegalStateException("Failed to add DockerSlaveSingle...");
+    }
+
+
+}

